backend/btrixcloud/pages.py

```diff
@@ -0,0 +1,467 @@
+"""crawl pages"""
+
+from datetime import datetime
+from typing import TYPE_CHECKING, Optional, Tuple, List, Dict, Any, Union
+from uuid import UUID, uuid4
+
+from fastapi import Depends, HTTPException
+import pymongo
+
+from .models import (
+    Page,
+    PageReviewUpdate,
+    PageQAUpdate,
+    Organization,
+    PaginatedResponse,
+    User,
+    PageNote,
+    PageNoteIn,
+    PageNoteEdit,
+    PageNoteDelete,
+)
+from .pagination import DEFAULT_PAGE_SIZE, paginated_format
+from .utils import from_k8s_date
+
+if TYPE_CHECKING:
+    from .crawls import CrawlOps
+    from .orgs import OrgOps
+    from .storages import StorageOps
+else:
+    CrawlOps = StorageOps = OrgOps = object
+
+
+# ============================================================================
+# pylint: disable=too-many-instance-attributes, too-many-arguments
+class PageOps:
+    """crawl pages"""
+
+    crawl_ops: CrawlOps
+    org_ops: OrgOps
+    storage_ops: StorageOps
+
+    def __init__(self, mdb, crawl_ops, org_ops, storage_ops):
+        self.pages = mdb["pages"]
+        self.crawl_ops = crawl_ops
+        self.org_ops = org_ops
+        self.storage_ops = storage_ops
+
+    async def add_crawl_pages_to_db_from_wacz(self, crawl_id: str):
+        """Add pages to database from WACZ files"""
+        try:
+            crawl = await self.crawl_ops.get_crawl(crawl_id, None)
+            org = await self.org_ops.get_org_by_id(crawl.oid)
+            wacz_files = await self.crawl_ops.get_wacz_files(crawl_id, org)
+            stream = await self.storage_ops.sync_stream_pages_from_wacz(org, wacz_files)
+            for page_dict in stream:
+                if not page_dict.get("url"):
+                    continue
+
+                await self.add_page_to_db(page_dict, crawl_id, crawl.oid)
+        # pylint: disable=broad-exception-caught, raise-missing-from
+        except Exception as err:
+            print(f"Error adding pages for crawl {crawl_id} to db: {err}", flush=True)
+
+    async def add_page_to_db(
+        self, page_dict: Dict[str, Any], crawl_id: str, oid: Optional[UUID] = None
+    ):
+        """Add page to database"""
+        page_id = page_dict.get("id")
+        if not page_id:
+            print(f'Page {page_dict.get("url")} has no id - assigning UUID', flush=True)
+            page_id = uuid4()
+
+        if await self.pages.find_one({"_id": page_id}):
+            return
+
+        if not oid:
```

Comment:
I think we can make this not optional, the `oid` should always be available.

Reply:
Done